rts/System/CMakeLists.txt

```diff
@@ -3,148 +3,185 @@ if (TRACY_PROFILE_MEMORY)
+	"${CMAKE_CURRENT_SOURCE_DIR}/Platform/Misc.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Platform/SharedLib.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Platform/ScopedFileLock.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Platform/SDL1_keysym.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Platform/Watchdog.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Platform/WindowManagerHelper.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Rectangle.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/SafeVector.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/SafeCStrings.c"
+	"${CMAKE_CURRENT_SOURCE_DIR}/SplashScreen.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/SpringApp.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/StartScriptGen.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Sync/DumpState.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Sync/FPUCheck.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Sync/Logger.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Sync/SHA512.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Sync/SyncChecker.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Sync/SyncDebugger.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Sync/SyncedFloat3.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Sync/backtrace.c"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Sync/get_executable_name.c"
+	"${CMAKE_CURRENT_SOURCE_DIR}/TdfParser.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Threading/ThreadPool.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/TimeProfiler.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/TimeUtil.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/Transform.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/UriParser.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/StringUtil.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/type2.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/float3.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/float4.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/SpringMath.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/SpringMem.cpp"
+	"${memoryProfileSource}"
+)
+
+add_library(system-creg STATIC)
+if (NOT(NO_CREG))
+	target_sources(system-creg PRIVATE
+		"${CMAKE_CURRENT_SOURCE_DIR}/creg/SerializeLuaState.cpp"
+		"${CMAKE_CURRENT_SOURCE_DIR}/creg/Serializer.cpp"
+		"${CMAKE_CURRENT_SOURCE_DIR}/creg/VarTypes.cpp"
+		"${CMAKE_CURRENT_SOURCE_DIR}/creg/creg.cpp"
+		"${CMAKE_CURRENT_SOURCE_DIR}/creg/creg_runtime_tests.cpp"
 	)
-make_global_var(sources_engine_System_Log_sinkOutputDebugString
-		"${CMAKE_CURRENT_SOURCE_DIR}/Log/OutputDebugStringSink.cpp"
+endif ()
+
+add_library(system-filesystem STATIC)
+target_sources(system-filesystem PRIVATE
+	"${CMAKE_CURRENT_SOURCE_DIR}/FileSystem/ArchiveNameResolver.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/FileSystem/ArchiveLoader.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/FileSystem/ArchiveScanner.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/FileSystem/CacheDir.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/FileSystem/DataDirLocater.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/FileSystem/DataDirsAccess.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/FileSystem/FileFilter.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/FileSystem/FileHandler.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/FileSystem/FileSystem.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/FileSystem/FileSystemAbstraction.cpp"
+	"${CMAKE_CURRENT_SOURCE_DIR}/FileSystem/FileSystemInitializer.cpp"
```

Comment:
For example this file, but many other, have ifdefs for dedicated, unitsync variants, for the problem I've described here: https://github.com/beyond-all-reason/spring/blob/master/rts/System/FileSystem/Archives/CMakeLists.txt#L20-L23

A single library target, can have AFAIK only a single set of defines, it's self contained, so, by putting it in a single target, you end up with a single variant: that won't work.

How do you plan to resolve this issue?

Reply:
At this point I cannot say anything final because I don't know, the concrete solution will crystalize with the progress of the work.
Currently, the idea is to create multiple targets with the same source but different definitions, likely with function helper to avoid boilerplate. This would be true with the state of the build system - the source _foo.cpp_ with definition _A_ contain different code then _foo.cpp_ with definition _B_. These should be different targets.